File: 0490_the_maze/has_path.java
// Time: O(m * n)
// Space: O(m * n)
class Solution {
    public boolean hasPath(int[][] maze, int[] start, int[] destination) {
        int[][] dirs = { { 0, 1 }, { 0, -1 }, { -1, 0 }, { 1, 0 } };
        Queue<int[]> queue = new LinkedList<>();
        queue.add(start);
        while (!queue.isEmpty()) {
            int[] curr = queue.poll();
            int x = curr[0], y = curr[1];
            if (x == destination[0] && y == destination[1]) {
                return true;
            }
            maze[x][y] = 2;
            for (int[] dir : dirs) {
                int r = x, c = y;
                while (isValidPath(maze, r + dir[0], c + dir[1])) {
                    r += dir[0];
                    c += dir[1];
                }
                if (maze[r][c] == 0) {
                    queue.add(new int[] { r, c });
                }
            }
        }
        return false;
    }

    private boolean isValidPath(int[][] maze, int row, int col) {
        return row >= 0 && col >= 0 && row < maze.length && col < maze[0].length && maze[row][col] != 1;
    }
}

File: 0490_the_maze/has_path.py
from collections import deque
from typing import List


def has_path(maze: List[List[int]], start: List[int], destination: List[int]) -> bool:
    """
    Time: O(n * m)
    Space: O(n * m)
    """
    n, m = len(maze), len(maze[0])
    dirs = [(0, 1), (1, 0), (-1, 0), (0, -1)]
    queue = deque([start])
    while queue:
        x, y = queue.popleft()
        if x == destination[0] and y == destination[1]:
            return True
        maze[x][y] = 2
        for i, j in dirs:
            r, c = x, y
            while n > r + i >= 0 and m > c + j >= 0 and maze[r + i][c + j] != 1:
                r, c = r + i, c + j
            if maze[r][c] == 0:
                queue.append([r, c])
    return False
